File: 11.py
import os

def load_books(file_path):
    """Загружает книги из файла."""
    books = []
    if os.path.exists(file_path):
        with open(file_path, 'r', encoding='utf-8') as file:
            for line in file:
                parts = line.strip().split()
                if len(parts) >= 5:
                    title = ' '.join(parts[:-4])
                    author, publisher, year, price = parts[-4:]
                    books.append((title, author, publisher, int(year), float(price)))
    return books

def save_books(file_path, books):
    """Сохраняет книги в файл."""
    with open(file_path, 'w', encoding='utf-8') as file:
        for book in books:
            file.write(f"{' '.join(map(str, book))}\n")

def display_books(books):
    """Отображает книги в табличном формате."""
    print(f"{'Название':<20} {'Автор':<15} {'Издательство':<15} {'Год':<5} {'Цена':<10}")
    print("-" * 70)
    for title, author, publisher, year, price in books:
        print(f"{title:<20} {author:<15} {publisher:<15} {year:<5} {price:<10.2f}")

def filter_books_by_price(books, min_price, max_price):
    """Фильтрует книги по диапазону цен."""
    return [book for book in books if min_price <= book[4] <= max_price]

def add_book(books):
    """Добавляет новую книгу."""
    title = input("Введите название книги: ")
    author = input("Введите автора книги: ")
    publisher = input("Введите издательство книги: ")
    year = int(input("Введите год издания книги: "))
    price = float(input("Введите цену книги: "))
    books.append((title, author, publisher, year, price))

def delete_book(books):
    """Удаляет книгу по индексу."""
    display_books(books)
    index = int(input("Введите индекс книги для удаления: "))
    if 0 <= index < len(books):
        books.pop(index)
    else:
        print("Некорректный индекс!")

def merge_books(file_path, books):
    """Добавляет книги из другого файла."""
    other_file = input("Введите путь к другому файлу: ")
    if os.path.exists(other_file):
        other_books = load_books(other_file)
        books.extend(other_books)
    else:
        print("Файл не найден!")

def main():
    file_path = input("Введите путь к файлу с книгами: ")
    books = load_books(file_path)

    while True:
        print("\nМеню:")
        print("1. Показать книги")
        print("2. Показать книги в диапазоне цен")
        print("3. Добавить книгу")
        print("4. Удалить книгу")
        print("5. Объединить с другим файлом")
        print("6. Сохранить и выйти")

        choice = input("Выберите действие: ")

        if choice == '1':
            display_books(books)
        elif choice == '2':
            min_price = float(input("Введите минимальную цену: "))
            max_price = float(input("Введите максимальную цену: "))
            filtered_books = filter_books_by_price(books, min_price, max_price)
            display_books(filtered_books)
        elif choice == '3':
            add_book(books)
        elif choice == '4':
            delete_book(books)
        elif choice == '5':
            merge_books(file_path, books)
        elif choice == '6':
            save_books(file_path, books)
            print("Данные сохранены. Выход из программы.")
            break
        else:
            print("Некорректный выбор!")

if __name__ == "__main__":
    main()

File: 11.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <iomanip>
#include <sstream>
#include <limits>
#include <regex> // Для работы с регулярными выражениями

using namespace std;

// Прототипы функций
struct Book; // Объявляем структуру Book

vector<Book> filter_books_by_price(const vector<Book>& books, float min_price, float max_price);
void add_book(vector<Book>& books);
void delete_book(vector<Book>& books);
void merge_books(const string& file_path, vector<Book>& books);

// Определение структуры
struct Book { // Структура Book представляет книгу и содержит поля для хранения названия, автора, издательства, года издания и цены.
    string title;
    string author;
    string publisher;
    int year;
    float price;
};

// Функция для чтения данных из файла
vector<Book> load_books(const string& file_path) {
    vector<Book> books;
    ifstream file(file_path);

    if (file.is_open()) {
        string line;
        int line_number = 0;
        regex pattern(R"((.+?)\s{2,}(.+?)\s{2,}(.+?)\s{2,}(\d{4})\s{2,}(\d+(\.\d+)?))");
        // Регулярное выражение для парсинга строки с двумя или более пробелами как разделителями

        while (getline(file, line)) {
            line_number++;
            smatch match;

            if (regex_match(line, match, pattern)) {
                Book book;
                book.title = match[1];
                book.author = match[2];
                book.publisher = match[3];
                book.year = stoi(match[4]);
                book.price = stof(match[5]);

                books.push_back(book);
            } else {
                cout << "Ошибка: Неверный формат данных в строке " << line_number << ": " << line << endl;
            }
        }

        file.close();
    } else {
        cout << "Ошибка: Не удалось открыть файл " << file_path << endl;
    }

    return books;
}
// Функция для сохранения данных в файл
void save_books(const string& file_path, const vector<Book>& books) {
    ofstream file(file_path);

    if (file.is_open()) {
        for (const auto& book : books) {
            file << book.title << "\t" << book.author << "\t" << book.publisher << "\t"
                 << book.year << "\t" << book.price << endl;
        }
        file.close();
    } else {
        cout << "Не удалось сохранить файл!" << endl;
    }
}

void display_books(const vector<Book>& books) {
    // Заголовки таблицы
    cout << left
         << setw(10) << "№\t"
         << setw(30) << "Название"
         << setw(30) << "Автор"
         << setw(20) << "Издательство\t"
         << setw(10) << "Год\t"
         << setw(10) << "Цена" << endl;

    cout << string(95, '-') << endl;  // Разделитель

    // Вывод данных о каждой книге с индексом
    for (size_t i = 0; i < books.size(); ++i) {
        cout << setw(10) << i  // Печатаем индекс
             << setw(30) << books[i].title
             << "\t"
             << setw(30) << books[i].author
             << "\t"
             << setw(20) << books[i].publisher
             << "\t"
             << setw(10) << books[i].year
             << fixed << setprecision(2) << setw(10) << books[i].price
             << endl;
    }
}

// Функция для фильтрации книг по диапазону цен
vector<Book> filter_books_by_price(const vector<Book>& books, float min_price, float max_price) {
    vector<Book> filtered_books;
    for (const auto& book : books) {
        if (book.price >= min_price && book.price <= max_price) {
            filtered_books.push_back(book);
        }
    }
    return filtered_books;
}

// Функция для добавления новой книги
void add_book(vector<Book>& books) {
    Book new_book;
    cout << "Введите название книги: ";
    getline(cin, new_book.title);
    cout << "Введите автора книги: ";
    getline(cin, new_book.author);
    cout << "Введите издательство книги: ";
    getline(cin, new_book.publisher);
    cout << "Введите год издания книги: ";
    cin >> new_book.year;
    cout << "Введите цену книги: ";
    cin >> new_book.price;
    cin.ignore();  // Для очистки буфера после ввода числа
    books.push_back(new_book);
}

// Функция для удаления книги по индексу
void delete_book(vector<Book>& books) {
    if (books.empty()) {
        cout << "Список книг пуст. Удаление невозможно.\n";
        return;
    }

    display_books(books);

    int index;
    cout << "Введите индекс книги для удаления: ";
    cin >> index;

    if (cin.fail()) {
        cin.clear();
        cin.ignore(numeric_limits<streamsize>::max(), '\n');
        cout << "Некорректный ввод! Попробуйте снова.\n";
        return;
    }

    if (index >= 0 && index < static_cast<int>(books.size())) {
        books.erase(books.begin() + index);
        cout << "Книга с индексом " << index << " удалена.\n";
    } else {
        cout << "Некорректный индекс! Попробуйте снова.\n";
    }
}


// Функция для добавления книг из другого файла
void merge_books(const string& file_path, vector<Book>& books) {
    string other_file_path;
    cout << "Введите путь к другому файлу: ";
    cin >> other_file_path;
    cin.ignore();  // Для очистки буфера после ввода пути

    vector<Book> other_books = load_books(other_file_path);
    books.insert(books.end(), other_books.begin(), other_books.end());
    cout << "Книги из другого файла успешно добавлены." << endl;
}

// Главная функция
int main() {
    string file_path;
    cout << "Введите путь к файлу с книгами: ";
    getline(cin, file_path);

    vector<Book> books = load_books(file_path);

    while (true) {
        cout << "\nМеню:\n";
        cout << "1. Показать все книги\n";
        cout << "2. Показать книги в диапазоне цен\n";
        cout << "3. Добавить книгу\n";
        cout << "4. Удалить книгу\n";
        cout << "5. Объединить с другим файлом\n";
        cout << "6. Сохранить и выйти\n";

        int choice;
        cout << "Выберите действие: ";
        cin >> choice;

        if (cin.fail()) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Некорректный ввод! Попробуйте снова.\n";
            continue;
        }

        cin.ignore(); // Очистка буфера
        if (choice == 1) {
            display_books(books);
        } else if (choice == 2) {
            float min_price, max_price;
            cout << "Введите минимальную цену: ";
            cin >> min_price;
            cout << "Введите максимальную цену: ";
            cin >> max_price;

            if (cin.fail()) {
                cin.clear();
                cin.ignore(numeric_limits<streamsize>::max(), '\n');
                cout << "Некорректный ввод! Попробуйте снова.\n";
                continue;
            }

            cin.ignore();
            vector<Book> filtered_books = filter_books_by_price(books, min_price, max_price);
            display_books(filtered_books);
        } else if (choice == 3) {
            add_book(books);
        } else if (choice == 4) {
            delete_book(books);
        } else if (choice == 5) {
            merge_books(file_path, books);
        } else if (choice == 6) {
            save_books(file_path, books);
            cout << "Данные сохранены. Выход из программы.\n";
            break;
        } else {
            cout << "Некорректный выбор! Попробуйте снова.\n";
        }
    }
    return 0;
}